src/java.base/share/classes/java/net/URLDecoder.java

```diff
@@ -187,61 +187,61 @@ public static String decode(String s, Charset charset) {
         while (i < numChars) {
             c = s.charAt(i);
             switch (c) {
-            case '+':
-                sb.append(' ');
-                i++;
-                needToChange = true;
-                break;
-            case '%':
-                /*
-                 * Starting with this instance of %, process all
-                 * consecutive substrings of the form %xy. Each
-                 * substring %xy will yield a byte. Convert all
-                 * consecutive  bytes obtained this way to whatever
-                 * character(s) they represent in the provided
-                 * encoding.
-                 */
-
-                try {
-
-                    // (numChars-i)/3 is an upper bound for the number
-                    // of remaining bytes
-                    if (bytes == null)
-                        bytes = new byte[(numChars-i)/3];
-                    int pos = 0;
-
-                    while ( ((i+2) < numChars) &&
-                            (c=='%')) {
-                        int v = Integer.parseInt(s, i + 1, i + 3, 16);
-                        if (v < 0)
+                case '+' -> {
+                    sb.append(' ');
+                    i++;
+                    needToChange = true;
+                }
+                case '%' -> {
+                    /*
+                     * Starting with this instance of %, process all
+                     * consecutive substrings of the form %xy. Each
+                     * substring %xy will yield a byte. Convert all
+                     * consecutive  bytes obtained this way to whatever
+                     * character(s) they represent in the provided
+                     * encoding.
+                     */
+
+                    try {
+
+                        // (numChars-i)/3 is an upper bound for the number
+                        // of remaining bytes
+                        if (bytes == null)
+                            bytes = new byte[(numChars - i) / 3];
+                        int pos = 0;
+
+                        while (((i + 2) < numChars) &&
+                                (c == '%')) {
+                            int v = Integer.parseInt(s, i + 1, i + 3, 16);
+                            if (v < 0)
+                                throw new IllegalArgumentException(
+                                        "URLDecoder: Illegal hex characters in escape "
+                                                + "(%) pattern - negative value");
+                            bytes[pos++] = (byte) v;
+                            i += 3;
+                            if (i < numChars)
+                                c = s.charAt(i);
+                        }
+
+                        // A trailing, incomplete byte encoding such as
+                        // "%x" will cause an exception to be thrown
+
```

Comment:
This change is a bit difficult to review and there doesn't appear much gain in changing from a regular switch to a switch expression in this case. Maybe we should just keep the regular switch here?

Reply:
OK. I've reverted change, as suggested. See 2f179b5